contracts/SolutionSubmitter.sol

```diff
@@ -0,0 +1,73 @@
+pragma solidity ^0.5.0;
+
+import "./BatchExchange.sol";
+import "@openzeppelin/contracts/ownership/Ownable.sol";
+
+interface GasToken {
+    function freeUpTo(uint256 value) external returns (uint256 freed);
+}
+
+/** @title SolutionSubmitter - A simple wrapper contract allowing to submit settlements
+ * to the BatchExchange contract while saving gas via the GST2 or CHI GasToken
+ */
+contract SolutionSubmitter is Ownable {
+    BatchExchange public exchange;
+    GasToken public gasToken;
+    uint256 public gasPriceThreshold;
+
+    constructor(
+        BatchExchange _exchange,
+        GasToken _gasToken,
+        uint256 _gasPriceThreshold
+    ) public {
+        exchange = _exchange;
+        setGasToken(_gasToken);
+        setGasPriceThreshold(_gasPriceThreshold);
+    }
+
+    function setGasToken(GasToken _gasToken) public onlyOwner {
+        gasToken = _gasToken;
+    }
+
+    function setGasPriceThreshold(uint256 _gasPriceThreshold) public onlyOwner {
+        gasPriceThreshold = _gasPriceThreshold;
+    }
+
+    /**
+     * Allow the owner to execute arbitrary code (e.g. transfer out solution fee rewards from BatchExchange)
+     */
+    function execute(address target, bytes calldata data) external onlyOwner returns (bool, bytes memory) {
+        // solium-disable-next-line security/no-low-level-calls
+        return target.call(data);
+    }
+
+    /**
+     * Wrapper around actual solution submission that uses gas tokens for discounts
+     */
+    function submitSolution(
+        uint32 batchId,
+        uint256 claimedObjectiveValue,
+        address[] memory owners,
+        uint16[] memory orderIds,
+        uint128[] memory buyVolumes,
+        uint128[] memory prices,
+        uint16[] memory tokenIdsForPrice
+    ) public gasDiscounted returns (uint256) {
+        return exchange.submitSolution(batchId, claimedObjectiveValue, owners, orderIds, buyVolumes, prices, tokenIdsForPrice);
+    }
+
+    /**
+     * Modifier to invoke original method and freeing up to half the maximum gas refund in gas tokens.
+     * Logic adjusted from https://1inch-exchange.medium.com/1inch-introduces-chi-gastoken-d0bd5bb0f92b
+     */
+    modifier gasDiscounted {
+        if (tx.gasprice >= gasPriceThreshold) {
+            uint256 gasStart = gasleft();
+            _;
+            uint256 gasSpent = 21000 + gasStart - gasleft() + (16 * msg.data.length);
+            gasToken.freeUpTo((gasSpent + 14154) / 41947);
```

Comment:
Could you add some comments about why these exact numbers are used? I don't understand where they come from :see_no_evil: .

Reply:
- 21000: cost of submitting any transaction
- gasStart - gasleft(): gas Used
- 16 * msg.data.length: must be 16 wei per byte sent as call data.

Just a quick guess.

Not sure about 14154 and 41947 though.

Reply:
Good point. I never questioned or investigated the computation as I took it from the official medium blog. It is however a good exercise in understanding what is happening and I will add a comment.

Reply:
I decided to simplify this calculation (only burn one token for every 48k gas spent). The original numbers where 14k fix cost for invoking `freeUpTo` (taken from https://gastoken.io/#comparison - search for "For the contract-based GasToken (GST2), we have")

The 41k was 48k - cost of burning a single token (~8k). However they are only approximations as well and given we cannot account for gas refunds coming from the solution submission itself (storage frees) slightly underestimating the token amount seems fine.